File: petties_mobile/lib/data/services/booking_service.dart
import 'package:flutter/material.dart';

import 'api_client.dart';
import '../models/booking.dart';
import '../models/clinic_service.dart';
import 'auth_service.dart';

/// BookingService - Handles booking-related API calls for mobile Staff
class BookingService {
  final ApiClient _apiClient = ApiClient.instance;

  /// Get booking by ID
  Future<BookingResponse> getBookingById(String bookingId) async {
    final response = await _apiClient.get('/bookings/$bookingId');
    return BookingResponse.fromJson(response.data);
  }

  /// Check-in booking (Staff action: CONFIRMED → IN_PROGRESS)
  /// Used for both clinic and home visits
  Future<BookingResponse> checkIn(String bookingId) async {
    final response = await _apiClient.post('/bookings/$bookingId/check-in');
    return BookingResponse.fromJson(response.data);
  }

  /// Checkout booking: thanh toán và hoàn tất lịch hẹn (IN_PROGRESS → COMPLETED)
  /// Chỉ gọi khi booking đang ở trạng thái IN_PROGRESS
  /// @param bookingId Booking UUID
  /// @param overriddenSosFee Optional: Override SOS fee (for staff adjustment)
  Future<BookingResponse> checkout(String bookingId,
      {double? overriddenSosFee}) async {
    final response = await _apiClient.post(
      '/bookings/$bookingId/checkout',
      data: overriddenSosFee != null
          ? {'overriddenSosFee': overriddenSosFee}
          : <String, dynamic>{},
    );
    return BookingResponse.fromJson(response.data);
  }

  /// Start moving to customer location (SOS/HOME_VISIT)
  /// Note: Status remains CONFIRMED or transitions to IN_PROGRESS based on backend logic
  Future<BookingResponse> startMoving(String bookingId) async {
    final response = await _apiClient.post('/bookings/$bookingId/start-moving');
    return BookingResponse.fromJson(response.data);
  }

  /// Staff arrived at customer location (SOS/HOME_VISIT)
  /// Keeps booking status IN_PROGRESS and marks arrival timestamp.
  Future<BookingResponse> arrived(String bookingId) async {
    final response = await _apiClient.post('/bookings/$bookingId/arrived');
    return BookingResponse.fromJson(response.data);
  }

  /// Notify pet owner that staff is on the way (HOME_VISIT/SOS only)
  /// Legacy: Just sends notification, use startMoving() for status change.
  /// @param bookingId Booking UUID
  Future<BookingResponse> notifyOnWay(String bookingId) async {
    final response =
        await _apiClient.post('/bookings/$bookingId/notify-on-way');
    return BookingResponse.fromJson(response.data);
  }

  /// Get staff home summary - optimized single API call for staff home screen
  /// Returns: today's booking count, pending count, in-progress count, and upcoming bookings
  Future<StaffHomeSummaryResponse> getStaffHomeSummary() async {
    final response = await _apiClient.get('/bookings/staff/home-summary');
    return StaffHomeSummaryResponse.fromJson(response.data);
  }

  /// Get my bookings (Pet Owner)
  Future<List<BookingResponse>> getMyBookings({
    String? status,
    int page = 0,
    int size = 10,
  }) async {
    final queryParams = {
      if (status != null) 'status': status,
      'page': page,
      'size': size,
    };

    final response = await _apiClient.get('/bookings/my-bookings',
        queryParameters: queryParams);

    if (response.data['content'] != null) {
      return (response.data['content'] as List)
          .map((json) => BookingResponse.fromJson(json))
          .toList();
    }
    return [];
  }

  /// Get my proxy bookings (Đặt hộ) - Pet Owner là người đặt thay
  Future<List<BookingResponse>> getMyProxyBookings({
    String? status,
    int page = 0,
    int size = 10,
  }) async {
    final queryParams = {
      if (status != null) 'status': status,
      'page': page,
      'size': size,
    };

    final response = await _apiClient.get(
      '/bookings/my/proxy',
      queryParameters: queryParams,
    );
    debugPrint('queryParams: $queryParams');
    debugPrint('response: ${response.data}');

    if (response.data['content'] != null) {
      return (response.data['content'] as List)
          .map((json) => BookingResponse.fromJson(json))
          .toList();
    }
    return [];
  }

  /// Cancel booking (Pet Owner)
  Future<BookingResponse> cancelBooking(String bookingId, String reason) async {
    final response = await _apiClient.post(
      '/bookings/$bookingId/cancel',
      queryParameters: {'reason': reason},
    );
    return BookingResponse.fromJson(response.data);
  }

  /// Get bookings by staff ID with filtering and pagination
  Future<Map<String, dynamic>> getBookingsByStaff({
    String? status,
    int page = 0,
    int size = 20,
  }) async {
    try {
      final user = await AuthService().getCurrentUser();

      final response = await _apiClient.get(
        '/bookings/staff/${user.userId}',
        queryParameters: {
          if (status != null && status != 'all') 'status': status,
          'page': page,
          'size': size,
          'sort': 'bookingDate,desc',
        },
      );
      return response.data;
    } catch (e) {
      rethrow;
    }
  }

  /// Remove service from booking (Only if isAddOn = true)
  Future<BookingResponse> removeServiceFromBooking(
      String bookingId, String serviceId) async {
    final response = await _apiClient.delete(
      '/bookings/$bookingId/services/$serviceId',
    );
    return BookingResponse.fromJson(response.data);
  }

  // ========== SHARED VISIBILITY ==========

  /// Get available services for add-on in a booking
  Future<List<ClinicServiceModel>> getAvailableServicesForAddOn(
      String bookingId) async {
    try {
      final response =
          await _apiClient.get('/bookings/$bookingId/available-add-ons');
      if (response.data is List) {
        return (response.data as List)
            .map((json) => ClinicServiceModel.fromJson(json))
            .toList();
      }
      return [];
    } catch (e) {
      rethrow;
    }
  }

  /// Add service to booking (Add-on service)
  Future<BookingResponse> addServiceToBooking(
      String bookingId, String serviceId) async {
    final response = await _apiClient.post(
      '/bookings/$bookingId/services',
      data: {'serviceId': serviceId},
    );
    return BookingResponse.fromJson(response.data);
  }

  /// Get all clinic bookings for today - Shared Visibility for Staff
  /// All staff in the clinic can see ALL bookings, with isMyAssignment flag
  /// to identify their own assignments.
  ///
  /// @param clinicId Clinic ID
  /// @returns List of ClinicTodayBooking with isMyAssignment flag
  Future<List<BookingResponse>> getClinicTodayBookings(String clinicId) async {
    try {
      final response = await _apiClient.get('/bookings/clinic/$clinicId/today');
      if (response.data is List) {
        return (response.data as List)
            .map((json) => BookingResponse.fromJson(json))
            .toList();
      }
      return [];
    } catch (e) {
      rethrow;
    }
  }
}

File: petties-web/src/services/bookingService.ts
/**
 * Booking API Service
 */
import axios from './api/client';
import type { Booking, CreateBookingRequest, ConfirmBookingRequest, AvailableStaffResponse, StaffAvailabilityCheckResponse, ConfirmBookingWithOptionsRequest } from '../types/booking';
import type { SosAlertMessage } from './websocket/sosWebSocket';
import type { BookingStatus } from '../types/booking';
import type { ClinicServiceResponse } from '../types/service';

// Spring Page response type
interface PageResponse<T> {
    content: T[];
    totalElements: number;
    totalPages: number;
    size: number;
    number: number;
}

// Staff option for dropdown selection
export interface StaffOption {
    staffId: string;
    fullName: string;
    avatarUrl?: string;
    specialty: string;
    specialtyLabel: string;
    isSuggested: boolean;
    bookingCount: number;
    hasAvailableSlots: boolean;
    availableServiceItemIds: string[];
    unavailableReason?: string;
}

const BOOKING_API = '/bookings';

/**
 * Get bookings by clinic (for Manager)
 */
export const getBookingsByClinic = async (
    clinicId: string,
    status?: BookingStatus,
    type?: string,
    page: number = 0,
    size: number = 20
): Promise<PageResponse<Booking>> => {
    const params = new URLSearchParams();
    params.append('page', page.toString());
    params.append('size', size.toString());
    if (status) params.append('status', status);
    if (type) params.append('type', type);

    const response = await axios.get(`${BOOKING_API}/clinic/${clinicId}?${params.toString()}`);
    return response.data;
};

/**
 * Get bookings by staff (for Staff)
 */
export const getBookingsByStaff = async (
    staffId: string,
    status?: BookingStatus,
    page: number = 0,
    size: number = 20
): Promise<PageResponse<Booking>> => {
    const params = new URLSearchParams();
    params.append('page', page.toString());
    params.append('size', size.toString());
    if (status) params.append('status', status);

    const url = `${BOOKING_API}/staff/${staffId}?${params.toString()}`;
    const response = await axios.get(url);
    return response.data;
};

/**
 * Get booking by ID
 */
export const getBookingById = async (bookingId: string): Promise<Booking> => {
    const response = await axios.get(`${BOOKING_API}/${bookingId}`);
    return response.data;
};

/**
 * Get booking by code
 */
export const getBookingByCode = async (bookingCode: string): Promise<Booking> => {
    const response = await axios.get(`${BOOKING_API}/code/${bookingCode}`);
    return response.data;
};

/**
 * Create a new booking (Pet owner)
 */
export const createBooking = async (request: CreateBookingRequest): Promise<Booking> => {
    const response = await axios.post(BOOKING_API, request);
    return response.data;
};

/**
 * Confirm booking and auto-assign staff (Manager)
 */
export const confirmBooking = async (
    bookingId: string,
    request?: ConfirmBookingRequest
): Promise<Booking> => {
    const response = await axios.post(`${BOOKING_API}/${bookingId}/confirm`, request || {});
    return response.data;
};

/**
 * Check staff availability before confirming booking
 * Returns detailed availability for each service and alternative time slot suggestions
 */
export const checkStaffAvailability = async (
    bookingId: string
): Promise<StaffAvailabilityCheckResponse> => {
    const response = await axios.get(`${BOOKING_API}/${bookingId}/availability`);
    return response.data;
};

/**
 * Confirm booking with options (Manager)
 * Supports allowPartial and removeUnavailableServices flags
 */
export const confirmBookingWithOptions = async (
    bookingId: string,
    request: ConfirmBookingWithOptionsRequest
): Promise<Booking> => {
    const response = await axios.post(`${BOOKING_API}/${bookingId}/confirm`, request);
    return response.data;
};

/**
 * Cancel booking
 */
export const cancelBooking = async (bookingId: string, reason: string): Promise<Booking> => {
    const response = await axios.post(`${BOOKING_API}/${bookingId}/cancel`, null, {
        params: { reason }
    });
    return response.data;
};

/**
 * Get available staff for reassigning a specific service
 */
export const getAvailableStaffForReassign = async (
    bookingId: string,
    serviceId: string
): Promise<AvailableStaffResponse[]> => {
    const response = await axios.get(`${BOOKING_API}/${bookingId}/services/${serviceId}/alternatives`);
    return response.data;
};

/**
 * Reassign staff for a specific service in a booking
 */
export const reassignStaffForService = async (
    bookingId: string,
    serviceId: string,
    newStaffId: string
): Promise<Booking> => {
    const response = await axios.post(`${BOOKING_API}/${bookingId}/services/${serviceId}/reassign`, {
        newStaffId
    });
    return response.data;
};

/**
 * Add a service to an active booking (IN_PROGRESS or CONFIRMED)
 * Used when staff wants to add extra services during home visit
 * Distance fee is NOT recalculated
 */
export const addServiceToBooking = async (
    bookingId: string,
    serviceId: string
): Promise<Booking> => {
    // Backend endpoint is /services (not /add-service)
    const response = await axios.post(`${BOOKING_API}/${bookingId}/services`, {
        serviceId
    });
    return response.data;
};

/**
 * Remove an add-on service from a booking
 */
export const removeServiceFromBooking = async (
    bookingId: string,
    bookingServiceId: string
): Promise<Booking> => {
    const response = await axios.delete(`${BOOKING_API}/${bookingId}/services/${bookingServiceId}`);
    return response.data;
};

/**
 * Get available services that can be added to this booking
 * Filters by specialty for Home Visit Staff
 */
export const getAvailableServicesForAddOn = async (
    bookingId: string
): Promise<ClinicServiceResponse[]> => {
    // Backend endpoint is /available-add-ons (not /available-services)
    const response = await axios.get(`${BOOKING_API}/${bookingId}/available-add-ons`);
    return response.data;
};

/**
 * Get available staff for confirming booking (dropdown selection)
 * Returns list of staff matching service specialty with availability status
 */
export const getAvailableStaffForConfirm = async (
    bookingId: string
): Promise<StaffOption[]> => {
    const response = await axios.get(`${BOOKING_API}/${bookingId}/available-staff-for-confirm`);
    return response.data;
};

/**
 * Check-in booking (Staff or Manager)
 * Transitions: CONFIRMED → IN_PROGRESS
 */
export const checkInBooking = async (bookingId: string): Promise<Booking> => {
    const response = await axios.post(`${BOOKING_API}/${bookingId}/check-in`);
    return response.data;
};

/**
 * Checkout booking (Staff/Manager action – thanh toán)
 * Chỉ gọi khi status IN_PROGRESS. Backend chuyển sang COMPLETED và xử lý thanh toán.
 * @param overriddenSosFee Optional: ghi đè phí SOS (cho booking SOS)
 */
export const checkoutBooking = async (
    bookingId: string,
    overriddenSosFee?: number | null
): Promise<Booking> => {
    const body = overriddenSosFee != null ? { overriddenSosFee } : {};
    const response = await axios.post(`${BOOKING_API}/${bookingId}/checkout`, body);
    return response.data;
};

// ========== SHARED VISIBILITY ==========

/**
 * Clinic Today Booking Response - extends Booking with isMyAssignment flag
 */
export interface ClinicTodayBooking extends Booking {
    isMyAssignment: boolean;
}

/**
 * Get all bookings for a clinic today - Shared Visibility for Staff
 * All staff in the clinic can see ALL bookings, with isMyAssignment flag
 * to identify their own assignments.
 *
 * @param clinicId Clinic ID
 * @returns List of ClinicTodayBooking with isMyAssignment flag
 */
export const getClinicTodayBookings = async (
    clinicId: string
): Promise<ClinicTodayBooking[]> => {
    const response = await axios.get(`${BOOKING_API}/clinic/${clinicId}/today`);
    return response.data;
};

// ========== SOS AUTO-MATCH ==========

const SOS_API = '/sos';

/**
 * Confirm SOS request (Clinic Manager)
 * The clinic agrees to handle this SOS booking
 */
export const confirmSosRequest = async (
    bookingId: string,
    assignedStaffId?: string
): Promise<Booking> => {
    const response = await axios.post(`${SOS_API}/${bookingId}/confirm`, {
        accepted: true,
        assignedStaffId
    });
    return response.data;
};

/**
 * Decline SOS request (Clinic Manager)
 * The clinic declines, system will try the next clinic
 * NOTE: Uses the same /confirm endpoint as confirmSosRequest.
 * Backend uses the `accepted` flag (true/false) to distinguish confirm vs decline.
 */
export const declineSosRequest = async (
    bookingId: string,
    reason?: string
): Promise<void> => {
    await axios.post(`${SOS_API}/${bookingId}/confirm`, {
        accepted: false,
        declineReason: reason
    });
};

/**
 * Get SOS matching status
 */
export const getSosMatchingStatus = async (bookingId: string): Promise<{
    status: string;
    clinicId?: string;
    clinicName?: string;
    message?: string;
}> => {
    const response = await axios.get(`${SOS_API}/${bookingId}/status`);
    return response.data;
};

/**
 * Get active SOS alerts for a clinic (Clinic Manager)
 * Used to sync alerts on mount (catch-up mechanism)
 */
export const getActiveSosAlerts = async (): Promise<SosAlertMessage[]> => {
    const response = await axios.get(`${SOS_API}/alerts`);
    return response.data;
};

// Named export for backwards compatibility and object-style imports
export const bookingService = {
    getBookingsByClinic,
    getBookingsByStaff,
    getBookingById,
    getBookingByCode,
    createBooking,
    confirmBooking,
    checkStaffAvailability,
    confirmBookingWithOptions,
    cancelBooking,
    getAvailableStaffForReassign,
    reassignStaffForService,
    addServiceToBooking,
    getAvailableServicesForAddOn,
    getAvailableStaffForConfirm,
    checkInBooking,
    checkoutBooking,
    removeServiceFromBooking,
    getClinicTodayBookings,
    confirmSosRequest,
    declineSosRequest,
    getSosMatchingStatus,
    getActiveSosAlerts,
};
